File: examples/javascript/02_tail_lookup.js
const BASE = process.env.JETNET_BASE_URL || "https://customer.jetnetconnect.com";
const EMAIL = process.env.JETNET_EMAIL;
const PASS = process.env.JETNET_PASSWORD;

if (!EMAIL || !PASS) {
  console.error("Set JETNET_EMAIL and JETNET_PASSWORD environment variables.");
  process.exit(1);
}

async function login() {
  const res = await fetch(`${BASE}/api/Admin/APILogin`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ emailAddress: EMAIL, password: PASS }),
  });
  if (!res.ok) throw new Error(`Login failed: HTTP ${res.status}`);
  const data = await res.json();
  return {
    bearerToken: data.bearerToken,
    apiToken: data.apiToken || data.securityToken,
  };
}

function isJetnetError(json) {
  return (
    json?.responsestatus &&
    String(json.responsestatus).toUpperCase().includes("ERROR")
  );
}

async function jetnetFetch(bearerToken, apiToken, path, opts = {}) {
  const url = `${BASE}${path}`.replaceAll("{apiToken}", apiToken);
  const res = await fetch(url, {
    ...opts,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${bearerToken}`,
      ...(opts.headers || {}),
    },
  });
  const json = await res.json();
  if (!res.ok || isJetnetError(json)) {
    throw new Error(
      isJetnetError(json) ? json.responsestatus : `HTTP ${res.status}`
    );
  }
  return json;
}

async function main() {
  const { bearerToken, apiToken } = await login();

  const tail = process.argv[2] || "N650GD";
  console.log(`Looking up tail: ${tail}\n`);

  const data = await jetnetFetch(
    bearerToken,
    apiToken,
    `/api/Aircraft/getRegNumber/${tail.toUpperCase()}/{apiToken}`,
    { method: "GET" }
  );

  const ac = data.aircraftresult;
  if (!ac || !ac.aircraftid) {
    console.log("Aircraft not found.");
    return;
  }

  console.log("Aircraft Details:");
  console.log(`  Aircraft ID : ${ac.aircraftid}`);
  console.log(`  Registration: ${ac.regnbr}`);
  console.log(`  Make        : ${ac.make}`);
  console.log(`  Model       : ${ac.model}`);
  console.log(`  Serial      : ${ac.serialnbr}`);
  console.log(`  Year Mfr    : ${ac.yearmfr}`);
  console.log(`  Base ICAO   : ${ac.baseicao || "N/A"}`);
  console.log(`  Base Airport: ${ac.baseairport || "N/A"}`);
  console.log(`  For Sale    : ${ac.forsale || "N/A"}`);

  const rels = ac.companyrelationships || [];
  if (rels.length > 0) {
    console.log("\nRelationships (flat schema from getRegNumber):");
    for (const r of rels) {
      console.log(
        `  ${r.companyrelation}: ${r.companyname} — ${r.contactfirstname || ""} ${r.contactlastname || ""} (${r.contacttitle || "N/A"})`
      );
    }
  }
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: examples/python/02_tail_lookup.py
"""
JETNET API -- Tail Number Lookup

Demonstrates how to look up an aircraft by registration (tail) number
and print key identification fields.

Required environment variables:
    JETNET_EMAIL    -- Your JETNET account email
    JETNET_PASSWORD -- Your JETNET account password
"""

import os
import sys
import requests

BASE = "https://customer.jetnetconnect.com"


def login(email, password):
    r = requests.post(
        f"{BASE}/api/Admin/APILogin",
        json={"emailAddress": email, "password": password},
    )
    r.raise_for_status()
    data = r.json()
    return data["bearerToken"], data["apiToken"]


def api(method, path, bearer, token, body=None):
    url = f"{BASE}{path}".replace("{apiToken}", token)
    headers = {
        "Authorization": f"Bearer {bearer}",
        "Content-Type": "application/json",
    }
    r = requests.request(method, url, headers=headers, json=body)
    r.raise_for_status()
    result = r.json()
    if "ERROR" in str(result.get("responsestatus", "")).upper():
        raise ValueError(f"JETNET error: {result['responsestatus']}")
    return result


def lookup_tail(bearer, token, tail_number):
    """Look up an aircraft by tail number. Returns the aircraftresult dict."""
    result = api("GET", f"/api/Aircraft/getRegNumber/{tail_number}/{{apiToken}}", bearer, token)
    return result.get("aircraftresult", {})


def main():
    email = os.environ.get("JETNET_EMAIL")
    password = os.environ.get("JETNET_PASSWORD")
    if not email or not password:
        print("Set JETNET_EMAIL and JETNET_PASSWORD environment variables.")
        sys.exit(1)

    tail = sys.argv[1] if len(sys.argv) > 1 else "N1KE"

    bearer, token = login(email, password)

    print(f"Looking up tail number: {tail}")
    ac = lookup_tail(bearer, token, tail.upper())

    if not ac.get("aircraftid"):
        print(f"Aircraft '{tail}' not found.")
        sys.exit(1)

    print(f"\n{'='*50}")
    print(f"  Aircraft ID:    {ac.get('aircraftid')}")
    print(f"  Registration:   {ac.get('regnbr')}")
    print(f"  Serial Number:  {ac.get('serialnbr')}")
    print(f"  Make:           {ac.get('make')}")
    print(f"  Model:          {ac.get('model')}")
    print(f"  Year Mfr:       {ac.get('yearmfr')}")
    print(f"  Year Delivered:  {ac.get('yeardlv')}")
    print(f"  Base ICAO:      {ac.get('baseicao')}")
    print(f"  Base Airport:   {ac.get('baseairport')}")
    print(f"  For Sale:       {ac.get('forsale')}")
    print(f"  Lifecycle:      {ac.get('lifecycle')}")
    print(f"{'='*50}")

    relationships = ac.get("companyrelationships", [])
    if relationships:
        print(f"\n  Company Relationships ({len(relationships)}):")
        for rel in relationships:
            print(f"    - {rel.get('companyrelation', 'N/A')}: {rel.get('companyname', 'N/A')}")
            contact = f"{rel.get('contactfirstname', '')} {rel.get('contactlastname', '')}".strip()
            if contact:
                print(f"      Contact: {contact} ({rel.get('contacttitle', '')})")


if __name__ == "__main__":
    main()
